concordium-consensus/src/Concordium/Scheduler/Environment.hs

```diff
@@ -488,40 +492,67 @@ class (StaticInformation m, ContractStateOperations m, MonadProtocolVersion m) =
   rejectingWith' !c reason = c >>= \case Right b -> return b
                                          Left a -> rejectTransaction (reason a)
 
+  -- |Add a contract upgrade to the 'ChangeSet'
+  addContractUpgrade :: ContractAddress
+                     -- ^The instance that should be upgraded.
+                     -> GSWasm.ModuleInterfaceA (InstrumentedModuleRef m GSWasm.V1)
+                     -- ^The set of receive names exposed by this new module for the instance.
+                     -> Set.Set GSWasm.ReceiveName
+                     -- ^The new module to use for execution.
+                     -> m ()
+
+
 -- |Index that keeps track of modifications of smart contracts inside a single
 -- transaction. This is used to cheaply detect whether a contract state has
 -- changed or not when a contract calls another.
 type ModificationIndex = Word
 
+-- |A pending update for v1 instances.
+data InstanceV1Update' mr = InstanceV1Update {
```

Comment:
Document the type parameter `mr`, and provide rationale for why it is a functor.